{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Install libraries for all code exercises in this lesson\n",
    "\n",
    "It takes about 8 to 9 minutes to install the libraries that are used in the exercises of this lesson.  To make your learning experience a bit smoother, please open this page in a new tab and run this code cell to install, then in your original page, continue onto the next video as the installation is running.  You should only have to install once for this lesson.  If you do \"menu\"->\"reset data\" at some point, you will have to reinstall the libraries again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple/\n",
      "Collecting zipline\n",
      "  Downloading https://pypi.tuna.tsinghua.edu.cn/packages/be/59/8c5802a7897c1095fdc409fb557f04df8f75c37174e80d2ba58c8d8a6488/zipline-1.3.0.tar.gz (2.5 MB)\n",
      "\u001b[K     |████████████████████████████████| 2.5 MB 140 kB/s eta 0:00:01     |████████▍                       | 645 kB 36 kB/s eta 0:00:50\n",
      "\u001b[31m    ERROR: Command errored out with exit status 1:\n",
      "     command: /root/anaconda3/envs/uda/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/tmp/pip-install-xlvydpg2/zipline/setup.py'\"'\"'; __file__='\"'\"'/tmp/pip-install-xlvydpg2/zipline/setup.py'\"'\"';f=getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__);code=f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' egg_info --egg-base /tmp/pip-install-xlvydpg2/zipline/pip-egg-info\n",
      "         cwd: /tmp/pip-install-xlvydpg2/zipline/\n",
      "    Complete output (82 lines):\n",
      "    ERROR: Exception:\n",
      "    Traceback (most recent call last):\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_vendor/urllib3/response.py\", line 425, in _error_catcher\n",
      "        yield\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_vendor/urllib3/response.py\", line 507, in read\n",
      "        data = self._fp.read(amt) if not fp_closed else b\"\"\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_vendor/cachecontrol/filewrapper.py\", line 62, in read\n",
      "        data = self.__fp.read(amt)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/http/client.py\", line 457, in read\n",
      "        n = self.readinto(b)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/http/client.py\", line 501, in readinto\n",
      "        n = self.fp.readinto(b)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/socket.py\", line 589, in readinto\n",
      "        return self._sock.recv_into(b)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/ssl.py\", line 1071, in recv_into\n",
      "        return self.read(nbytes, buffer)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/ssl.py\", line 929, in read\n",
      "        return self._sslobj.read(len, buffer)\n",
      "    socket.timeout: The read operation timed out\n",
      "    \n",
      "    During handling of the above exception, another exception occurred:\n",
      "    \n",
      "    Traceback (most recent call last):\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/cli/base_command.py\", line 186, in _main\n",
      "        status = self.run(options, args)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/commands/wheel.py\", line 161, in run\n",
      "        resolver.resolve(requirement_set)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/legacy_resolve.py\", line 177, in resolve\n",
      "        discovered_reqs.extend(self._resolve_one(requirement_set, req))\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/legacy_resolve.py\", line 333, in _resolve_one\n",
      "        abstract_dist = self._get_abstract_dist_for(req_to_install)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/legacy_resolve.py\", line 282, in _get_abstract_dist_for\n",
      "        abstract_dist = self.preparer.prepare_linked_requirement(req)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/operations/prepare.py\", line 482, in prepare_linked_requirement\n",
      "        hashes=hashes,\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/operations/prepare.py\", line 287, in unpack_url\n",
      "        hashes=hashes,\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/operations/prepare.py\", line 159, in unpack_http_url\n",
      "        link, downloader, temp_dir.path, hashes\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/operations/prepare.py\", line 303, in _download_http_url\n",
      "        for chunk in download.chunks:\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_internal/network/utils.py\", line 39, in response_chunks\n",
      "        decode_content=False,\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_vendor/urllib3/response.py\", line 564, in stream\n",
      "        data = self.read(amt=amt, decode_content=decode_content)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_vendor/urllib3/response.py\", line 529, in read\n",
      "        raise IncompleteRead(self._fp_bytes_read, self.length_remaining)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/contextlib.py\", line 130, in __exit__\n",
      "        self.gen.throw(type, value, traceback)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pip/_vendor/urllib3/response.py\", line 430, in _error_catcher\n",
      "        raise ReadTimeoutError(self._pool, None, \"Read timed out.\")\n",
      "    pip._vendor.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out.\n",
      "    Traceback (most recent call last):\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/setuptools/installer.py\", line 128, in fetch_build_egg\n",
      "        subprocess.check_call(cmd)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/subprocess.py\", line 363, in check_call\n",
      "        raise CalledProcessError(retcode, cmd)\n",
      "    subprocess.CalledProcessError: Command '['/root/anaconda3/envs/uda/bin/python3', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmpbs4fa21s', '--quiet', 'numpy>=1.11.1']' returned non-zero exit status 2.\n",
      "    \n",
      "    During handling of the above exception, another exception occurred:\n",
      "    \n",
      "    Traceback (most recent call last):\n",
      "      File \"<string>\", line 1, in <module>\n",
      "      File \"/tmp/pip-install-xlvydpg2/zipline/setup.py\", line 323, in <module>\n",
      "        **conditional_arguments\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/setuptools/__init__.py\", line 143, in setup\n",
      "        _install_setup_requires(attrs)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/setuptools/__init__.py\", line 138, in _install_setup_requires\n",
      "        dist.fetch_build_eggs(dist.setup_requires)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/setuptools/dist.py\", line 698, in fetch_build_eggs\n",
      "        replace_conflicting=True,\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pkg_resources/__init__.py\", line 783, in resolve\n",
      "        replace_conflicting=replace_conflicting\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pkg_resources/__init__.py\", line 1066, in best_match\n",
      "        return self.obtain(req, installer)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/pkg_resources/__init__.py\", line 1078, in obtain\n",
      "        return installer(requirement)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/setuptools/dist.py\", line 754, in fetch_build_egg\n",
      "        return fetch_build_egg(self, req)\n",
      "      File \"/root/anaconda3/envs/uda/lib/python3.7/site-packages/setuptools/installer.py\", line 130, in fetch_build_egg\n",
      "        raise DistutilsError(str(e))\n",
      "    distutils.errors.DistutilsError: Command '['/root/anaconda3/envs/uda/bin/python3', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmpbs4fa21s', '--quiet', 'numpy>=1.11.1']' returned non-zero exit status 2.\n",
      "    ----------------------------------------\u001b[0m\n",
      "\u001b[31mERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.\u001b[0m\n",
      "\u001b[?25h"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "#!{sys.executable} -m pip install -i  https://pypi.tuna.tsinghua.edu.cn/simple/ -r requirements.txt\n",
    "!{sys.executable} -m pip install -i  https://pypi.tuna.tsinghua.edu.cn/simple/ zipline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "uda",
   "language": "python",
   "name": "uda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
